src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManagerImpl.java

```diff
@@ -0,0 +1,227 @@
+     */
+    Transaction open() {
+        currentDepth++;
+
+        if (stack.size() == currentDepth) {
+            stack.add(new TransactionImpl(currentDepth));
+        }
+
+        TransactionImpl current = stack.get(currentDepth);
+        current.lifecycle = Transaction.Lifecycle.OPEN;
+        return current;
+    }
+
+    void validateCurrentThread() {
+        if (Thread.currentThread() != thread) {
+            String errorMessage = String.format(
+                    "Attempted to access transaction state from thread %s, but this transaction is only valid on thread %s.",
+                    Thread.currentThread().getName(),
+                    thread.getName());
+            throw new IllegalStateException(errorMessage);
+        }
+    }
+
+    public Transaction.Lifecycle getLifecycle() {
+        if (currentDepth == -1) {
+            return Transaction.Lifecycle.NONE;
+        } else {
+            return stack.get(currentDepth).lifecycle;
+        }
+    }
+
+    private class TransactionImpl implements Transaction {
+        final int nestingDepth;
+        final ArrayList<CloseCallback> closeCallbacks = new ArrayList<>();
+        Lifecycle lifecycle = Lifecycle.NONE;
+
+        TransactionImpl(int nestingDepth) {
+            this.nestingDepth = nestingDepth;
+        }
+
+        void validateCurrentTransaction() {
+            validateCurrentThread();
+
+            if (currentDepth == -1 || stack.get(currentDepth) != this) {
+                String errorMessage = String.format(
+                        "Transaction function was called on a transaction with depth %d, but the current transaction has depth %d.",
+                        nestingDepth,
+                        currentDepth);
+                throw new IllegalStateException(errorMessage);
+            }
+        }
+
+        // Validate that this transaction is open.
+        private void validateOpen() {
+            if (lifecycle != Lifecycle.OPEN) {
+                throw new IllegalStateException("Transaction operation cannot be applied to a closed transaction.");
+            }
+        }
+
+        @Override
+        public Transaction openNested() {
+            validateCurrentTransaction();
+            validateOpen();
+            return open();
+        }
+
+        private void close(Result result) {
+            validateCurrentTransaction();
+            validateOpen();
+            // Block transaction operations
+            lifecycle = Lifecycle.CLOSING;
+
+            // Note: it is important that we don't let exceptions corrupt the global state of the transaction manager.
+            // That is why any callback has to run inside a try block.
+            RuntimeException closeException = null;
+
+            // Invoke callbacks in reverse order
+            for (int i = closeCallbacks.size()-1; i >= 0; i--) {
+                try {
+                    closeCallbacks.get(i).onClose(this, result);
```

Comment:
I am fine doing it either way, but given we don't explicitly need the index, do we want to instead make use of `ListIterator` and just iterate going backwards?

Reply:
So the idea here was to avoid allocations entirely. (That includes the iterator). I doubt that the list iterator would be any fast, but I'm not sure it would make a practical difference.